### PR TITLE
JAMES-1752 Handle multipart mediatype when retrieving text/html body

### PR DESCRIPTION
**Pull request by aduprat**

### CODE REVIEW

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java [hunk 1]
```diff
@@ -368,4 +368,91 @@ public void getMessagesShouldReturnAttachmentsWhenSome() throws Exception {
...
+            .contentType(ContentType.JSON)
+            .header("Authorization", accessToken.serialize())
+            .body("[[\"getMessages\", {\"ids\": [\"" + username + "|inbox|1\"]}, \"#0\"]]")
+        .when()
+            .post("/jmap")
+        .then()
+            .statusCode(200)
+            .body(NAME, equalTo("messages"))
+            .body(ARGUMENTS + ".list", hasSize(1))
+            .body(ARGUMENTS + ".list[0].hasAttachment", equalTo(true))
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: you should use a CONSTANT for list[0]

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java [hunk 2]
```diff
@@ -368,4 +368,91 @@ public void getMessagesShouldReturnAttachmentsWhenSome() throws Exception {
...
+    public void getMessagesShouldReturnTextBodyWhenSomeAttachmentsAndTextmail() throws Exception {
+        jmapServer.serverProbe().createMailbox(MailboxConstants.USER_NAMESPACE, username, "inbox");
+
+        ZonedDateTime dateTime = ZonedDateTime.parse("2014-10-30T14:12:00Z");
+        jmapServer.serverProbe().appendMessage(username, new MailboxPath(MailboxConstants.USER_NAMESPACE, username, "inbox"),
+                ClassLoader.getSystemResourceAsStream("twoAttachmentsTextPlain.eml"), Date.from(dateTime.toInstant()), false, new Flags());
+        
+        await();
+        
+        String expectedBody = "html\n";
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: very confusing (:

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/htmlAndTextMultipartWithOneAttachment.eml [hunk 3]
```diff
@@ -0,0 +1,1389 @@
...
+--------------172F9470CFA3BF3417835D92--
+
+--------------7AF1D14DE1DFA16229726B54
+Content-Type: image/png;
+ name="Capture du 2016-03-03 11:44:22.png"
+Content-Transfer-Encoding: base64
+Content-Disposition: attachment;
+ filename="Capture du 2016-03-03 11:44:22.png"
+
+iVBORw0KGgoAAAANSUhEUgAABOAAAAKSCAYAAAB7gaSAAAAABHNCSVQICAgIfAhkiAAAABl0
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/htmlAndTextMultipartWithOneAttachment.eml]: could you use a smaller attachment please ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/message/IndexableMessage.java [hunk 4]
```diff
@@ -89,6 +98,15 @@ private void copyMessageFields(MailboxMessage message, ZoneId zoneId) {
         this.properties = message.getProperties();
     }
 
+    private String valueFromMimePart(Optional<MimePart> mimePart, 
+            Function<MimePart, Optional<String>> functionToValue,
+            String defaultValue) {
+        return mimePart
+                .map(functionToValue)
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/message/IndexableMessage.java]: flatmap ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/message/MimePart.java [hunk 5]
```diff
@@ -260,4 +261,23 @@ public HeaderCollection getHeaderCollection() {
                 .flatMap((mimePart) -> Stream.concat(Stream.of(mimePart), mimePart.getAttachmentsStream()));
     }
 
+    private static class MimePartComparator implements Comparator<MimePart> {
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/message/MimePart.java]: Come on, I have to read the whole class to understand what it is supposed to do ...
You don't even have full unit test coverage, it's not ok

**Comment by mbaechler** [comment]
The tests are ok but everything else is way too complex, sorry, I don't understand half of what you changed.
I may be a bit dumb, but you probably too complex code.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java [hunk 6]
```diff
@@ -368,4 +368,91 @@ public void getMessagesShouldReturnAttachmentsWhenSome() throws Exception {
...
+            .body(ARGUMENTS + ".list", hasSize(1))
+            .body(ARGUMENTS + ".list[0].hasAttachment", equalTo(true))
+            .body(ATTACHMENTS, hasSize(2))
+            .body(ARGUMENTS + ".list[0].preview", equalTo(expectedBody))
+            .body(ARGUMENTS + ".list[0].textBody", equalTo(expectedBody))
+            .body(ARGUMENTS + ".list[0].htmlBody", nullValue());
+    }
+
+    @Test
+    public void getMessagesShouldReturnHTMLBodyWhenOneAttachmentAndBothHTMLAndTextmail() throws Exception {
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: Why ? What prevent us from having also the text body ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java [hunk 7]
```diff
@@ -368,4 +368,91 @@ public void getMessagesShouldReturnAttachmentsWhenSome() throws Exception {
...
+
+    @Test
+    public void getMessagesShouldReturnTextBodyWhenSomeAttachmentsAndTextmail() throws Exception {
+        jmapServer.serverProbe().createMailbox(MailboxConstants.USER_NAMESPACE, username, "inbox");
+
+        ZonedDateTime dateTime = ZonedDateTime.parse("2014-10-30T14:12:00Z");
+        jmapServer.serverProbe().appendMessage(username, new MailboxPath(MailboxConstants.USER_NAMESPACE, username, "inbox"),
+                ClassLoader.getSystemResourceAsStream("twoAttachmentsTextPlain.eml"), Date.from(dateTime.toInstant()), false, new Flags());
+        
+        await();
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: No need for await() as you are not checking the index

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java [hunk 8]
```diff
@@ -368,4 +368,91 @@ public void getMessagesShouldReturnAttachmentsWhenSome() throws Exception {
...
+
+    @Test
+    public void getMessagesShouldReturnHTMLBodyWhenSomeAttachmentsAndHTMLmail() throws Exception {
+        jmapServer.serverProbe().createMailbox(MailboxConstants.USER_NAMESPACE, username, "inbox");
+
+        ZonedDateTime dateTime = ZonedDateTime.parse("2014-10-30T14:12:00Z");
+        jmapServer.serverProbe().appendMessage(username, new MailboxPath(MailboxConstants.USER_NAMESPACE, username, "inbox"),
+                ClassLoader.getSystemResourceAsStream("twoAttachments.eml"), Date.from(dateTime.toInstant()), false, new Flags());
+        
+        await();
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: No need for await() as you are not reading from ES after this call.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java [hunk 9]
```diff
@@ -368,4 +368,91 @@ public void getMessagesShouldReturnAttachmentsWhenSome() throws Exception {
...
+
+    @Test
+    public void getMessagesShouldReturnHTMLBodyWhenOneAttachmentAndBothHTMLAndTextmail() throws Exception {
+        jmapServer.serverProbe().createMailbox(MailboxConstants.USER_NAMESPACE, username, "inbox");
+
+        ZonedDateTime dateTime = ZonedDateTime.parse("2014-10-30T14:12:00Z");
+        jmapServer.serverProbe().appendMessage(username, new MailboxPath(MailboxConstants.USER_NAMESPACE, username, "inbox"),
+                ClassLoader.getSystemResourceAsStream("htmlAndTextMultipartWithOneAttachment.eml"), Date.from(dateTime.toInstant()), false, new Flags());
+        
+        await();
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: Again, you do not need this await...

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java (hunk 1 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java (hunk 2 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/htmlAndTextMultipartWithOneAttachment.eml (hunk 3 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/htmlAndTextMultipartWithOneAttachment.eml]: OK

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/message/IndexableMessage.java (hunk 4 above)
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/message/IndexableMessage.java]: Removed

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/message/MimePart.java (hunk 5 above)
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/message/MimePart.java]: Removed

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java (hunk 6 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java (hunk 8 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java (hunk 9 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java (hunk 7 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: OK

**Comment by aduprat** [comment]
#275

**Pull request closed without merging**